legend-engine-xts-graphQL/legend-engine-xt-graphQL-query/src/main/java/org/finos/legend/engine/query/graphQL/api/execute/GraphQLExecute.java

```diff
@@ -232,12 +248,50 @@ private Response executeGraphQLQuery(String queryClassPath, String mappingPath,
             if (isQueryIntrospection(graphQLQuery))
             {
                 pureModel = modelLoader.call();
-                org.finos.legend.pure.m3.coreinstance.meta.pure.metamodel.type.Class<?> _class = pureModel.getClass(queryClassPath);
-                org.finos.legend.pure.generated.Root_meta_external_query_graphQL_metamodel_sdl_Document queryDoc = toPureModel(document, pureModel);
+                return executeIntrospection(queryClassPath, document, pureModel);
+            }
+            else
+            {
+                if (graphQLPlanCache != null)
+                {
+                    planWithSerialized = graphQLPlanCache.getIfPresent(graphQLCacheKey);
+                    if (planWithSerialized == null) //cache miss, generate the plan and add to the cache
+                    {
+                        LOGGER.debug(new LogInfo(profiles, LoggingEventType.GRAPHQL_EXECUTE, "Cache miss. Generating new plan").toString());
+                        pureModel = modelLoader.call();
+                        planWithSerialized = buildPlanWithParameterUsingDataspace(queryClassPath, dataspacePath, executionContext, document, graphQLQuery, pureModel, graphQLCacheKey);
+                        graphQLPlanCache.put(graphQLCacheKey, planWithSerialized);
+                    }
+                    else
+                    {
+                        LOGGER.debug(new LogInfo(profiles, LoggingEventType.GRAPHQL_EXECUTE, "Cache hit. Using previously cached plan").toString());
+                    }
+                }
+                else   //no cache so we generate the plan
+                {
+                    pureModel = modelLoader.call();
+                    planWithSerialized = buildPlanWithParameterUsingDataspace(queryClassPath, dataspacePath, executionContext, document, graphQLQuery, pureModel, graphQLCacheKey);
+                }
+            }
+        }
+        catch (Exception e)
+        {
+            return ExceptionTool.exceptionManager(e, LoggingEventType.EXECUTE_INTERACTIVE_ERROR, profiles);
+        }
+        return execute(profiles, planWithSerialized, graphQLQuery);
+    }
 
-                return Response.ok("{" +
-                        "  \"data\":" + core_external_query_graphql_transformation_transformation_introspection_query.Root_meta_external_query_graphQL_transformation_introspection_graphQLIntrospectionQuery_Class_1__Document_1__String_1_(_class, queryDoc, pureModel.getExecutionSupport()) +
-                        "}").type(MediaType.TEXT_HTML_TYPE).build();
+    private Response executeGraphQLQuery(String queryClassPath, String mappingPath, String runtimePath, Document document, GraphQLCacheKey graphQLCacheKey, MutableList<CommonProfile> profiles, Callable<PureModel> modelLoader)
+    {
+        List<SerializedNamedPlans> planWithSerialized;
```

Comment:
All this code is almost same as above method. As I understand, only type of cache key (or what it holds changes). Should modularize

Reply:
Done